drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/graal/compiler/OCLNodeLIRBuilder.java

```diff
@@ -683,6 +683,26 @@ public void visitEndNode(final AbstractEndNode end) {
             return;
         }
 
+        // Move the phi assignment outside the loop.
+        // Only do that for the phi nodes that are not inside "else { break; }" blocks.
+        Block curBlock = (Block) gen.getCurrentBlock();
+        boolean shouldSkipStmt = false;
+        if (curBlock.getBeginNode() instanceof LoopExitNode) {
+            LoopExitNode loopExitNode = (LoopExitNode) curBlock.getBeginNode();
+            LoopBeginNode loopBeginNode = loopExitNode.loopBegin();
+            Block loopBeginBlock = loopBeginNode.graph().getLastSchedule().getNodeToBlockMap().get(loopBeginNode);
+            for (Block pred : curBlock.getPredecessors()) {
+                if (pred == loopBeginBlock) {
+                    shouldSkipStmt = true;
+                    break;
+                }
+            }
+        }
+
+        if (shouldSkipStmt) {
+            append(new OCLLIRStmt.BeginSkipStmt());
```

Comment:
So these instructions are just to keep control flow?

Reply:
Yes, if phi values are used before, inside, and after a loop, then we need to make sure that after the loop, the phi variables have been assigned (even if the loop did not iterate)